### ISSUE TITLE
Update List values on the browser.

### ISSUE THREAD

**Issue by Ajitweb** [dsuryd/dotNetify#19]
Hi Dicky,

In my module with master and child records, I am in the process of doing CRUD with the child records. I will not save it to database as of now. I am only gathering and validating records of child table and will save to the database only after the master fields are also validated.

I have a potran_list with the child records related to the master and use the foreach to build a table in the browser.

`<tbody id="tbodytablepotranlist" style="height:200px" data-bind="foreach: { data: potran_list, as: 'record' }, vmItemKey: 'sr'">
`

Each row also has a button, that will open up a modal with the values of that row for the user to edit. (These values are stored in separate observable properties and is not related to the list)

If the user chooses to save the changes, then he clicks the save button on the modal which is bound to a method (potranbtnok) which updates the specific list's record with the property values from the modal.

```
        public bool potranbtnok
        {
            get { return false; }
            set
            {
                potran_oktosave = true;

                //validate the properties - todo
                if (potran_oktosave)
                {
                    //Update potran_list
                    var recset = potran_list.FirstOrDefault(x => x.sr == potran_sr);
                    recset.name = potran_name;  //Store the value from the modal into recset which will also update potran_list.
                    //Do the same for other properties

                    Potranlineitemmodalclose = "close the modal";  //This will activate the vm.$on script which closes the modal
                }
            }
        }

```
When the potran_list is updated, I want to show the user the updated values in the table. How to "push" the list changes of that specific record to the browser.?

```
Also, I tried removing and adding it like this
`
this.RemoveList(() => potran_list, potran_sr);`
this.AddList(() => potran_list, recset);`
```
and this removes the record from the browser but does not add it back.

But I can always add a new record with a new sr,

`this.AddList(() => potran_list, new potran_listViewModel {sr=3,, code = potran_code, name=potran_name, confirmdocno = potran_confirmdocno, qty = potran_qty, price = potran_price, total = potran_total });
`

I looked into your Simple list example - but don't know how to apply to the above situation. 
Need your help. Thanks.

**Comment by dsuryd** [comment]
I don't have it documented by oversight, but there's the `UpdateList` method to do exactly what you need.  Example usage: 
`this.UpdateList( () => MyList, new ListItem { Id = 1, Value = "new value" } );`

**Comment by Ajitweb** [comment]
Thanks Dicky. The UpdateList worked like a charm.

About, ICommand - I just had a glance - but could not figure out the advantage over the regular button method. Is it the parameters that can be passed to the button actions?

**Comment by dsuryd** [comment]
It's just a pattern commonly used in WPF apps.  No worries if you don't use it.

**Comment by Ajitweb** [comment]
Ok. Thanks Dicky.

**Comment by Ajitweb** [comment]
Hi Dicky,
Another question regarding UpdateList()
In a server method, can I issue UpdateList() - only once?
When I try to update it more than one time then all the previous changes get reverted back. Only the latest record's UpdateList() is shown on the browser. 

I did the following test
```
var recset = potran_list.FirstOrDefault(x => x.sr == 1);
recset.name = "first";
this.UpdateList(() => potran_list, recset);
recset = null;
recset = potran_list.FirstOrDefault(x => x.sr == 2);
recset.name = "second";
this.UpdateList(() => potran_list, recset);
recset = null;
recset = potran_list.FirstOrDefault(x => x.sr == 3);
recset.name = "third";
this.UpdateList(() => potran_list, recset);
recset = null;
recset = potran_list.FirstOrDefault(x => x.sr == 4);
recset.name = "fourth";
this.UpdateList(() => potran_list, recset);
recset = null;
recset = potran_list.FirstOrDefault(x => x.sr == 5);
recset.name = "fifth";
this.UpdateList(() => potran_list, recset);
recset = null;
```
On the browser, I can see only last record's name as 'fifth'. All the previous record's values gets reverted back to its original on the browser. But If I see the potran_list on the server, the names have changed for all records.

**Comment by dsuryd** [comment]
I think you're asking for batch update, which is not currently supported.  The workaround would be to call `PushUpdates `for every `UpdateList `call.

**Comment by Ajitweb** [comment]
Awesome! Thanks once again.

**Comment by Ajitweb** [comment]
Hi Dicky,
I am not opening a new thread as this question may not have anything to do with Dotnetify.

In one of the properties on the server in the set - i am generating a pdf file to be send back to the browser.
I traditional MVC app, I used the Response object to send it back but here the Response is not being recognized by the compiler maybe because I am inheriting from BaseVM instead of Controller class.

How do I send a file back to the browser?

Thanks

**Comment by dsuryd** [comment]
When the pdf is ready, set a VM property with the pdf file location.  On the client side, listen to this property changed event, and then execute an AJAX request to fetch the pdf and display it.

**Comment by Ajitweb** [comment]
Thanks Dicky.

I did that - works perfectly, and instead of using AJAX to fetch the pdf, I simply issued a 
`window.open(pdffilenameonserver);`
in the vm.$on script

I tested it on IE as well as Chrome and it works great. 

Still, do you find downloading via AJAX is better in some way (limitations in window.open)?

There is this HTML5 way of dowloading shown here
https://muaz-khan.blogspot.com/2012/10/save-files-on-disk-using-javascript-or.html

Also another jquery plugin for downloading is found here
http://johnculviner.com/jquery-file-download-plugin-for-ajax-like-feature-rich-file-downloads/
But requires some sort of cookie writing of which I am not very clear on where to incorporate in my code.
HttpContext.Current.Response.SetCookie(new HttpCookie("fileDownload", "true") { Path = "/" });

Thanks once again

Regards

Ajit

**Comment by dsuryd** [comment]
I think it's perfectly fine to use window.open.  The reason you want to use AJAX is if you want to control the placement of the PDF content inside the existing page, or do advanced handling like progress bar and error notification.   

If you want to do the approach using those plugins, it has to be done outside of  dotNetify; basically your javascript establishes regular HTTP call to your ASP.NET controller API where you will have access to the HttpContext.

**Comment by Ajitweb** [comment]
Thanks Dicky for your valuable input - as always. As of now I am using dotNetify to inform the browser when the server is busy preparing the pdf. Then it is just the time taken to download that pdf where the user will be clueless as to what is happening. Fortunately, as of now in my routines, the pdf file size is very small so it is ok. But I will have to definitely incorporate some kind of progress bar for larger pdfs in the future. 

Thanks once again.

**Comment by dsuryd** [comment]
Yes, don't use dotNetify/signalR to pass files.  Your steps involving the Controller API are correct.  You probably want to push progress feedback while the pdfs are being uploaded, so the trick is how to have the code in the API controller communicates the progress to the view model so it can be pushed back to the client.  This will involve initially passing signalR connection id through the HTTP Request data to the Controller and then get the instance of the view model through DotNetify.VMControllerFactory.GetInstance API.  If you're ready to implement this and need more details, just ping me again.

**Comment by Ajitweb** [comment]
Hi Dicky. Yes, I will get in touch with you, when I start my Document upload programs.
In the meantime, I was trying to find out a Knockout custom binder that will help to create table. There are many good ones out there - but for some reason or the other - it never worked in my scenario (especially when I am trying to show the table in a boostrap modal). Issues like double header, erratic column widths etc with the existing ones forced me write (and learn) another binder.

[Here](https://jsbin.com/jehufey/edit?html,js,output) it is - in case someone has the same issues as me and needs a table with search, sort, export to csv and print.

**Comment by dsuryd** [comment]
That looks awesome!  Did you pass the entire 'tableko' config from the back-end VM, or only the list data?

**Comment by Ajitweb** [comment]
Hello Dicky,

For the jsbin example, I have populated the table data in the js.

But in my app, the table data comes from the property defined in your dotnetify's VM.
I have defined all my custom ko binders in the appRequire.js under the

```
require(['jquery', 'knockout'.....], function ($, ko) {
    $(function () {
.....

```
I don't know, if that is the correct location, but when I try putting it as a separate js file, it ends up in errors. Anyway I am glad you like what I have done so far. - Long way to go as my existing desktop ERP app (written in Microsoft Visual Foxpro) is huge and I shudder when I think of converting - some of the complex logic that I have written there - to Dotnet. As of now, I am concentrating to convert my ERP rather than looking to use new technologies like dotnet core.

Thanks to you for Dotnetify with its SignalR mechanism - it is the closest I could get to mimic a desktop programming environment. Otherwise, I would have gone the dreaded ajaxy way of programming (Also,I simply hate the Get and Post method of programming.). With dotnetify the backend logic is almost taken care of - rest was to tackle the verbose html - but that too is somewhat simplified by using custom knockout binders.
